### ISSUE TITLE
Proposal: Make `Window` inherit from DependencyObject

### ISSUE THREAD

**Issue by dotMorten** [microsoft/microsoft-ui-xaml#7305]
# Proposal: The Window class should inherit from DependencyObject

## Summary
One of the most annoying things about the Window class today, and one that diverges a lot from other XAML frameworks is that the Window class is not a DependencyObject, even though we can declare it in xaml.
It limits a lot of typical use-cases.

## Rationale
- Window properties does not support binding, as there's no property bag to store it in.
- Window does not support attached properties to extend how it behave right from xaml (like setting backdrop with helper classes, titlebar settings, manage window stage across app sessions etc).
- You cannot find the Window that a UI Element belongs to, because the root XAML element has the Parent as 'null' instead of Window.
- You cannot use resources to set properties on the window (like theme resources applied to titlebar via helper classes in attached properties).

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to set Window properties through binding | Must |
| This proposal will allow developers to extend Window behavior in XAML with attached properties | Must |
| This proposal will allow developers to reach to theme and resource changes at the Window level | Must |
| This proposal will allow end users to discover which window a given UI element is part of | Must |

## Important Notes
- Window today doesn't have a baseclass, so injecting a baseclass (at least from a .NET point of view) is not a breaking change.
      Before: `public class Window : ICustomQueryInterface, IWinRTObject, IDynamicInterfaceCastable, IEquatable<Window>`
      After: `public class Window : DependencyObject, ICustomQueryInterface, IWinRTObject, IDynamicInterfaceCastable, IEquatable<Window>`

**Comment by mpettit2253** [comment]
Isn't this the job of the Content property of the Window (same as in UWP?)

**Comment by dotMorten** [comment]
@mpettit2253 I'm not following? Of the items listed above, none of these are set on the Content.

**Comment by mpettit2253** [comment]
I'm speaking from a fair lack of experience here, so if I get things really wrong I apologize.  I had meant that the Window class in WinUI3 and the AppWindow class in the AppSDK aren't meant to be UIElements or even Visuals like in WPF.  They seem to be lightweight bridges to the underlying HWND infrastructure.  Choosing to expose them as a name in XAML isn't so helpful, though their recommendation noted below might work.

If I am misreading this and you were able to do these things in Windows 8/UWP, let me know. I am sad about certain things in those stacks that have been left behind like Suspend/Resume.  Those features allowed me to write a multi-instance ArcGIS Runtime app that used SceneView and MapView for the main view.  Users could open 10 or more instances to work with disparate but related locations, and only the non-minimized ones would take memory, etc.  Your resume magic worked great.

That being said, I'll try using the current recommendations in the docs to see if the functionality you list can be done without super pain.  It looks like the two big recommendations are:
1) if you need XAML Visual derived features, put a panel or a page as the root and go from there:
https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/guides/winui3#visual-state-manager-and-pageresources

2) the changing ability to flow into the title bar area may be compatible with their above recommendation:
https://docs.microsoft.com/en-us/windows/apps/develop/title-bar?tabs=winui3

**Comment by JesseCol** [comment]
This could also potentially help #9960.  (thanks @lhak)

**Comment by dotMorten** [comment]
I've been thinking a lot about this issue I logged lately, and I might have changed my mind on it. Here's the push-back on the rationales I initially logged, and curious if anyone disagrees:
>  Window properties does not support binding, as there's no property bag to store it in.

x:Bind is fully supported. Considering x:Bind is better for several reasons, performance and type-safety being one, but also AoT support etc, I'm not sure it really needs to support classic binding.

> Window does not support attached properties to extend how it behave right from xaml (like setting backdrop with helper classes, titlebar settings, manage window stage across app sessions etc).

You could subclass Window and add these properties, thereby get the same feature, but with type-safety. While perhaps not as convenient, it's definitely not a blocker and I struggle with coming up with good scenarios for attached properties on the Window.

> You cannot find the Window that a UI Element belongs to, because the root XAML element has the Parent as 'null' instead of Window.

This is covered in https://github.com/microsoft/microsoft-ui-xaml/issues/9857 already. I also think changing parent to suddenly be a Window instance might be a breaking behavior, and Parent is defined on `FrameworkElement`, not `DependencyObject`. While changing baseclass to `FrameworkElement` could be an alternative option, it would introduce a large amount of members that doesn't really make sense on Window.

> You cannot use resources to set properties on the window (like theme resources applied to titlebar via helper classes in attached properties).

While I see this argument to some extent, since I logged this issue we got a new `TitleBar` control which will live in the UI tree and you get all those features. There's still a matter of controlling the [`AppWindowTitleBar`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindowtitlebar?view=windows-app-sdk-2.0), but I think that should be solved in a more first-class way or via a x:bindable custom property on Window. Also AppWindowTitleBar lives as a subproperty under AppWindow, so feels a bit out of scope for what is needed here.

What do others think? Does this issue really make sense?

**Comment by HO-COOH** [comment]
@dotMorten 
1. Classic `Binding` is still necessary in case of a delayed binding when `x:Bind` causes nullptr exception. Scenarios like a visualstate trigger binding to a control's property.
2.
> You could subclass Window and add these properties, thereby get the same feature, but with type-safety

I don't understand how subclassing a window provides type-safety. You only get a raw pointer which can be `reinterpret_cast` to any type. Absolute destroying type-safety.

3. The new `TitleBar` control does not solve the question like: **Why can't I simply use a binding to change the `Window.Title` when the property is right there? Why do I need to use a new Control?**

Making `Window` to be a `DependencyObject` is perfectly valid for all the reasons that otherwise requires awkward workarounds or fix-ups, or a new control to workaround.

**Comment by YexuanXiao** [comment]
In my understanding, DependencyObject is used not only to support Binding, but also to support other XAML features. The fact that Window does not inherit from DependencyObject causes considerable inconsistency, and all of this is difficult to document. When people write code, they are usually unaware of these issues, and it is also difficult to trace the root cause back to this, resulting in a large amount of buggy code.

**Comment by dotMorten** [comment]
@HO-COOH Wrt binding Title, what I don't understand is why you can't change `Title="{Binding AppTitle}"` to  `Title="{x:Binding AppTitle, Mode=OneWay}"` ?

> Classic Binding is still necessary in case of a delayed binding when x:Bind causes nullptr exception.

Wouldn't that be a bug in your viewmodel though? Should it ever return null for something that shouldn't be null?

> Scenarios like a visualstate trigger binding to a control's property.

Do you have some examples where you'd want to use visual state to act on a window? 

> I don't understand how subclassing a window provides type-safety.

If I have a property like `public string Width {get; set;}` and I try to bind that to the Width of my Window, I'll get compile time errors telling me about the type mismatch. A classic binding wouldn't do that and instead cause problems at runtime. The same goes for attached properties, vs adding a new property to my subclass Window. ie `<Window MyCustomStringProperty="{x:Bind Title}" />` vs `<Window MyCustomClass.MyCustomAttachedStringProperty="{Binding Title}" />`

> In my understanding, DependencyObject is used not only to support Binding, but also to support other XAML features.

@YexuanXiao Which features are you referring to? For reference these are the APIs you'd get access to:

<img width="658" height="384" alt="Image" src="https://github.com/user-attachments/assets/6cc252ed-9e52-4a69-bbf4-38fc60370c3c" />

**Comment by dotMorten** [comment]
One more thing I can't believe I overlooked after trying this out in a quick prototype: Inheriting from DependencyObject does _not_ give you binding support on Window, for the very simple obvious reason that there is no DataContext property. That is defined on FrameworkElement, The problematic bit is changing the inheritance to FrameworkElement (or Control), brings in a huge set of APIs that might just not make sense for a Window:
<img width="5467" height="12468" alt="Image" src="https://github.com/user-attachments/assets/45bb8ee8-edf1-4de9-b154-1ddced4eacd9" />

**Comment by YexuanXiao** [comment]
If Window inherits from DependencyObject or FrameworkElement to make it more consistent with other Xaml components, then it should inherit from them; otherwise, it should not be allowed to be used in conjunction with Xaml. The WinUI project template creates MainWindow.xaml, which severely amplifies the problem. I oppose maintaining the status quo; in any case, there should be a solution.

**Comment by hexawyz** [comment]
In WPF, Window inherits from `ContentControl`, so I wonder if it would be possible to do something similar in the case of WinUI.
It is true that there is some kind of weirdness in the situation of WPF `Window` as it acts as both the frame and some content element that is the client area of the window, so I understand the desire by WinUI to break away from this weirdness not "cleanly" solvable by OOP. But in the end, it seems like it creates more problems than it solves.

It might be more desirable to just let `Window` expose itself as `ContentControl` like in WPF. After all, the client area of the Window can rightfully be considered to be a content control. I'm not even sure it is impossible to do that change without paying some perf cost.

The main problem, and where the cost lies, is that considering Window to be a `FrameworkElement`/`Control`/`ContentControl` (whichever you chose, but let's go to the utmost and pay the price for `ContentControl` does introduce a new indirection, at least in the logical sense, as there can now be a template for the `Window`, which means that `Window.Content` is no longer the actual `Root` (well, kinda, but not in the Visual Tree sense) because Window now has a Background and all.

But I don't think that's a problem that is impossible to overcome.
First of all, it might be fine to pay the cost of this new default template that may be composed of a border and an ContentPresenter. If by default everything is transparent, there will be nothing to be seen there.
Then, it might very well be possible to allow the Window to shortcut most of the complex logic when the default template is used (which should, effectively, be the default and most frequent case) and expose the exact same visual tree as the current-day `Window`. I feel like this would address most compatibility or performance issues that could arise with pre-existing code.
This solution does however come with a slight cost in (in)consistency, as a "default" `Window`'s Visual Tree would be different than the one for a `Window` whose `Background` has been defined, for example.

Then, the last problem, is that you are now implicitly allowing to place `Window` as the content of other elements… But it seems like it should be relatively easy to force this operation to always fail, as a `Window` can be considered the same as a `Control` already having a logical parent.

In the end, I'm even thinking it wouldn't be that difficult to come up with a `Window2` prototype deriving from `ContentControl`… After all, nothing should forbid a control instance from creating its host `Window` itself, right?

**Comment by HO-COOH** [comment]
> In WPF, Window inherits from ContentControl, so I wonder if it would be possible to do something similar in the case of WinUI.

Of course it's possible. I already [implemented one](https://github.com/microsoft/microsoft-ui-xaml/discussions/11536).

**Comment by hexawyz** [comment]
> Of course it's possible. I already https://github.com/microsoft/microsoft-ui-xaml/discussions/11536.

Wow nice, you're saving me from taking a shot at it myself ! 😄

**Comment by HO-COOH** [comment]
@dotMorten 
1. The binding issues I was talking about: #4966, #11472 
2. It is impossible to use  `x:Uid` to localize the `Title` on a Window. See #10928

These can be worked around by using `Binding` or function bindings or whatever. But, a workaround IS a workaround. They are unintuitive (otherwise there would be no issue report at the first place) and is documented nowhere.

3. The visual state scenario I was talking about is, sometimes you can have different window sizes under different "state" (because of different control state might take up different amount of spaces). If window size can be used as a VisualState.Setter's Target, that would be convenient. However, the harder-blocker is that, under a `Window`, you can't define `VisualState` anywhere that's only targeting controls, not even Window. You have to put it in a `Page`.
4. Also, consider animated window size in my discussion I linked above

**Comment by dotMorten** [comment]
@HO-COOH for the converter issue, why do you consider function binding a workaround? It's much less code than a converter, avoids casts, gives you type safety and better performance. IMHO it's the better solution for "modern binding" over converters.
Good point on the x:Uid, but I wonder if that can't just be fixed separately.

> Of course it's possible. I already https://github.com/microsoft/microsoft-ui-xaml/discussions/11536.

That approach is interesting with it owning the window inside it, and I've done something similar with full control template support.
Speaking of what is possible, I just had an agent work on changing the inheritance to `DependencyObject`, `FrameworkElement` and `Control` to play with the approaches. All of it is _possible_ (albeit it's likely still riddled with problems and edge-cases as it was just a quick and dirty test), but that doesn't mean it's necessarily a good idea. Yes inheriting from Control gives you control templates/styling which is rather nice, but I'm not loving all the stuff it inherits from its base classes. It adds a lot of irrelevant stuff to Window. It also has a breaking behavior effect on what the XamlRoot is, which might make it a no-go.

If we only inherit from `DependencyObject`, you'll get a few of the benefits, but there's no DataContext (`FrameworkElement` introduces that), so a new `DataContext` property would have to be introduced (again that seems possible to do but it feels odd to do all that when x:Bind can do the property bindings and is the better way to support AoT going forward). 

If you want to attach a `VisualStateManager` to the Window, it has to at least be a `FrameworkElement`.

> In the end, I'm even thinking it wouldn't be that difficult to come up with a Window2

Yes you can definitely do that, as @HO-COOH linked to his approach and I've done a similar thing myself. I just don't like having to have a new Window type, as it means any existing code written to manage or work with Window wouldn't work, not to mention it's confusing to have two window types in the framework. It might be better to just have something like this in a 3rd party extension library or as a helper class in your own project.


Bottom line is I think the proposal **as stated above** to just inherit from `DependencyObject` doesn't actually add much value.

**Comment by hexawyz** [comment]
> Bottom line is I think the proposal as stated above to just inherit from `DependencyObject` doesn't actually add much value.

Since we seem to be a few to agree, why not update the proposal to `ContentControl` ?

> but I'm not loving all the stuff it inherits from its base classes. It adds a lot of irrelevant stuff to Window.

I do agree, but don't you think the benefits outweigh the downsides?

> It also has a breaking behavior effect on what the XamlRoot is, which might make it a no-go.

Seems like this could be handled with some compatibility feature in the worst case?
Otherwise, accepting this small breaking change in a major release might be enough?

**Comment by dotMorten** [comment]
>  why not update the proposal to ContentControl 

I stated my concerns above regarding that already. Whether it's Control, FrameworkElement or ContentControl doesn't really matter too much. My same concern applies. You asked if the benefits outweight the downsides. I'd be worried all these non-functional properties lead to more confusion, and there's a huge amount of risk involved in trying to override the layout system's existing behavior for Window. Not impossible, but it might be too risky to be reasonably feasible. In my prototype I already hit lots and lots of weird behavior issues, and quite a lot of code required to make this change and handle all the special cases.

I'm still not 100% clear what it is Binding can do that you can't do with x:Bind today. Especially considering not using x:Bind prevents a lot of the AoT support. When I logged this issue back in the day, I wasn't too concerned about AoT because it was barely working, but now I avoid Bindings are all cost, and if x:Bind can do it all, then it kind of negates the reasoning for this issue as it was originally logged.

I'm of course more than happy to be challenged on this - that's what the thread is here for.

**Comment by YexuanXiao** [comment]
@dotMorten Could you make your experimental branch public and share the difficulties you encountered during the implementation? For example, open a draft PR marked as [do not merge] for people to review, so that everyone can understand the actual situation.

**Comment by dotMorten** [comment]
@YexuanXiao They are completely 100% vibe coded, and not fully baked so I'd rather not share. Just local commits I never pushed. I did them mostly to try and play with the resulting API. You can use an agent to do a similar thing.
